### PR TITLE
Add steerable wheels

### PR DESCRIPTION
**Pull request by mutability**
This follows on from #14185 which laid the groundwork for vehicles rotating about points other than the origin.

It adds a basic algorithm for picking a rotation center based on finding the point that minimizes the resistance to turning from the non-steering wheels. Current gotchas: (1) currently _everything_ is a non-steering wheel; (2) I've made the model's implementation verbose rather than fast.

Looking for some feedback on 

(a) is this the right model for finding the rotation center?
(b) what's the best way to get steerable wheels in without breaking all existing vehicles?

### CODE REVIEW

**Comment by mutability** [comment]
For adding steerable wheels, one approach would be to add a new steerable part type for each wheel type.

Then what do we do about existing vehicles? On load, try to pick some wheels to convert to steerable versions?

**Comment by Rivet-the-Zombie** [comment]
>Then what do we do about existing vehicles? On load, try to pick some wheels to convert to steerable versions?

Nothing too difficult; if we add a new category of wheel for steering, we'd have to go through the vehicle .jsons and update them.

**Comment by mutability** [comment]
I meant more about existing-in-a-save-game vehicles; ideally I don't really want to break them all..

**Comment by Rivet-the-Zombie** [comment]
Sorry, I misunderstood what you meant.

**Comment by kevingranade** [comment]
The only thing I was able to come up with as far as converting wheels is to
find the frontmost row of wheels and convert them to steering wheels.

Alternately if we have a reasonable fallback for handling vehicles with no
steering wheels (pivot point at the average location of all wheels?), we
can just leave them using the old wheel types and it's up to the player to
fix them up.  This would mean we have to maintain support for that fallback
though.

**Comment by mutability** [comment]
It'd be nice to have "you have no steering wheels" mean "you have no steering" (otherwise there's no real reason to put any on) which means we can't easily have a fallback. Turning the front row of wheels into steerable wheels sounds reasonable. We can use the existing standard vehicles as testcases.

**Comment by mutability** [comment]
OK, I've fleshed this out a bit now - how's it look?

**Comment by Coolthulhu** [comment]
What about boats and other moded wheel parts?

**Comment by mutability** [comment]
I have not looked at the modded vehicles at all yet.

**Comment by Griffinhart** [comment]
I'd say convert all wheels to steering wheels, and then players themselves can choose whether they want (A|F|R)WD. 

This assumes that steering wheels aren't particularly rare, though.

**Comment by kevingranade** [comment]
That also assumes we want to hand out AWD for 'free'.  If we want to
differentiate based on transmission type (and I do), it makes that more
difficult.

**Comment by Griffinhart** [comment]
How does having lots of AWD vehicles (in saves predating this PR) make differentiating based on transmission type more difficult?

**Comment by kevingranade** [comment]
If the default is AWD, then when actual AWD is added, it will effectively
nerf every vehicle made after the change.

**Comment by Griffinhart** [comment]
Oh, I thought meant "differentiating" in like... a "sort and categorize" sense of the word.

>it will effectively nerf every vehicle

Uh, assuming that AWD is strictly worse than both (F|R)WD, which to my understanding (I'm not a car person) isn't true.

And either way, assuming *any* kind of default transmission will muck up the distribution of transmission types.

Though I guess it doesn't matter too much, if steerable wheels are approximately as common as non-steerable wheels, or if converting between the two incurs no significant additional cost.

(I suggested default AWD because going from AWD -> not AWD is more convenient for a player than going from not AWD -> AWD if steerable wheels are rare or otherwise inconvenient to acquire/install.)

**Comment by mutability** [comment]
The current implementation does this:

 * Saved vehicles get converted to FWD
 * Almost all vehicle prototypes are FWD, from memory only the bubble car is AWD
 * Steerable wheels don't require a different item, they just use regular wheels; changing which wheels are steerable only requires the tools/charges/skill needed to remove and re-add the parts.
 * Adding >1 steerable axle (nb: axle, not wheel) has an additional skill requirement (similar to what happens with engines) so if you want AWD on a 10-axle monstrosity you will need a fairly high mechanics skill.

which I think already covers most of the comments above.

**Comment by mutability** [comment]
(I guess this is more "front wheel steering" not "front wheel drive")

**Comment by Griffinhart** [comment]
>Almost all vehicle prototypes are FWD, from memory only the bubble car is AWD

The Humvee should be AWD (IRL ones are).

**Comment by mutability** [comment]
The military humvee is certainly 4WD, but does it have all-wheel steering? A quick look at the specs didn't turn it up.

**Comment by Griffinhart** [comment]
Ahhh, hm. I wasn't aware there was a difference! (Not a car guy.)

A quick check on Wikipedia https://en.wikipedia.org/wiki/Steering#Active_four-wheel_steering seems to indicate that quite a few sports cars do though, so maybe the Sports Car in CDDA should have all-wheel steering.

**Comment by mutability** [comment]
Possibly some of the bigger trucks too, though there are no really long trucks in the prototypes.

**Comment by mutability** [comment]
More vehicles have all-wheel steering now.
Added a TRACKED flag to deal with blazemod's tracks/treads.
Updated the tankmod vehicles.
Boats should be handled now.

**Comment by mutability** [comment]
I think this is ready to go now.

File: src/veh_interact.cpp [hunk 1]
```diff
@@ -558,6 +577,13 @@ bool veh_interact::can_install_part(int msg_width){
                             dif_eng);
     }
 
+    if (dif_steering > 0) {
+        engine_string = string_format(
```
Comment (kevingranade) [inline src/veh_interact.cpp]: Pretty sure you meant this to be ```steering_string = ...```

File: src/game.cpp [hunk 2]
```diff
@@ -11659,13 +11659,29 @@ void game::pldrive(int x, int y)
         }
     }
 
+    int turn_delta = 15 * x;
+    if (turn_delta != 0) {
```
Comment (kevingranade) [inline src/game.cpp]: Style nit, project style is ```if( turn_delta != 0 ) {```.
The same with the rest of the diff.

File: src/veh_interact.cpp (hunk 1 above)
Comment (mutability) [inline src/veh_interact.cpp]: Right, yeah. Actually we could probably get away with one less var here if we never have "steerable engines"

File: src/game.cpp (hunk 2 above)
Comment (mutability) [inline src/game.cpp]: I'm fighting a decade's worth of muscle memory here; a makefile target to restyle the code to the right form would help a lot.

Comment (kevingranade) [inline src/game.cpp]: I'm working on getting enough code compliant with the current style to do
that, but it's an uphill battle.

**Pull request merged**